data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/PeerForwardingProcessorDecorator.java

```diff
@@ -15,38 +15,67 @@
 
 import java.util.Collection;
 import java.util.HashSet;
+import java.util.List;
 import java.util.Set;
+import java.util.stream.Collectors;
 
 public class PeerForwardingProcessorDecorator implements Processor<Record<Event>, Record<Event>> {
-    private final Processor innerProcessor;
     private final PeerForwarder peerForwarder;
-    private final String pluginId;
-    private final Set<String> identificationKeys;
+    private final Processor innerProcessor;
 
-    public PeerForwardingProcessorDecorator(final Processor innerProcessor,
-                                            final PeerForwarderProvider peerForwarderProvider,
-                                            final String pipelineName,
-                                            final String pluginId) {
-        this.innerProcessor = innerProcessor;
-        this.pluginId = pluginId;
+    public static List<Processor> decorateProcessors(
+            final List<Processor> processors,
+            final PeerForwarderProvider peerForwarderProvider,
+            final String pipelineName,
+            final String pluginId) {
 
-        if (innerProcessor instanceof RequiresPeerForwarding) {
-            identificationKeys = new HashSet<> (((RequiresPeerForwarding) innerProcessor).getIdentificationKeys());
-        } else {
-            throw new UnsupportedPeerForwarderPluginException("Peer Forwarding is only supported for plugins which implement RequiresPeerForwarding interface.");
+        Set<String> identificationKeys;
+        final Processor firstInnerProcessor = processors.get(0);
```

Comment:
This should check the size first. It is probably perfectly legitimate to return an empty list if the input list is empty. This scenario could have a test case as well.